comp/apm/etwtracer/impl/apmetwtracerimpl.go

```diff
@@ -0,0 +1,334 @@
+		if err == io.EOF {
+			return
+		}
+
+		if err != nil {
+			a.log.Errorf("Read error: %v", err)
+			return
+		}
+
+		switch h.CommandResponse {
+		case registerCommand:
+			a.log.Infof("Registering process with ID %d", pid)
+			err = a.AddPID(pid)
+			if err != nil {
+				a.log.Errorf("Failed to reconfigure the ETW provider for PID %d: %v", pid, err)
+				h.CommandResponse = errorResponse
+			} else {
+				h.CommandResponse = okResponse
+			}
+		case unregisterCommand:
+			a.log.Infof("Unregistering process with ID %d", pid)
+			err = a.RemovePID(pid)
+			if err != nil {
+				a.log.Errorf("Failed to reconfigure the ETW provider for PID %d: %v", pid, err)
+				h.CommandResponse = errorResponse
+			} else {
+				h.CommandResponse = okResponse
+			}
+		default:
+			a.log.Infof("Unsupported command %d", h.CommandResponse)
+		}
+		h.Size = headerSize
+
+		err = binary.Write(c, binary.LittleEndian, &h)
+
+		// Client disconnected
+		if err == io.EOF {
+			return
+		}
+
+		if err != nil {
+			a.log.Errorf("Read error: %v", err)
+			return
+		}
+	}
+}
+
+func (a *apmetwtracerimpl) start(_ context.Context) error {
+	a.log.Infof("Starting Datadog APM ETW tracer component")
+	var err error
+	etwSessionName := "Datadog APM ETW tracer"
+	a.session, err = a.etw.NewSession(etwSessionName)
+	if err != nil {
+		a.log.Errorf("Failed to create the ETW session '%s': %v", etwSessionName, err)
+		// Don't fail the Agent startup
+		return nil
+	}
+
+	a.pipeListener, err = winio.ListenPipe(serverNamedPipePath, &winio.PipeConfig{
+		MessageMode: true,
+	})
+	if err != nil {
+		a.log.Errorf("Failed to listen to named pipe '%s': %v", serverNamedPipePath, err)
+		// Don't fail the Agent startup
+		return nil
+	}
+	go func() {
+		for {
+			conn, err := a.pipeListener.Accept()
+			if err != nil {
+				// net.ErrClosed is returned when pipeListener is Close()'d
+				if err != net.ErrClosed {
+					a.log.Warnf("Could not accept new client:", err)
+				}
+				return
+			}
+			go a.handleConnection(conn)
+		}
+	}()
+
+	go func() {
+		// StartTracing blocks the caller
+		_ = a.session.StartTracing(func(e *etw.DDEventRecord) {
+			a.log.Infof("Received event %d for PID %d", e.EventHeader.EventDescriptor.ID, e.EventHeader.ProcessID)
+			pid := uint64(e.EventHeader.ProcessID)
+			var pidCtx pidContext
+			var ok bool
+			if pidCtx, ok = a.pids[pid]; !ok {
+				// We may still be receiving events a few moments
+				// after a process un-registers itself, no need to log anything here.
+				return
+			}
+			pidCtx.lastSeen = time.Now()
+			ev := clrEvent{
+				header: header{
+					Magic:           a.magic,
+					CommandResponse: clrEventResponse,
+				},
+				EventHeader:    e.EventHeader,
+				UserData:       etwutil.GoBytes(unsafe.Pointer(e.UserData), int(e.UserDataLength)),
+				UserDataLength: e.UserDataLength,
+			}
+			ev.header.Size = uint16(unsafe.Sizeof(ev)) + e.UserDataLength
+			_, writeErr := pidCtx.conn.Write(etwutil.GoBytes(unsafe.Pointer(&ev), int(ev.header.Size)))
```

Comment:
What if the pipe is broken we will keep pollution our log. Can we check based on error it we should try to get the pipe again or remove tracking maybe.